File: web/src/api/okr.ts
import { api } from "./client";

export type OKRKeyResult = {
  id: string;
  objective_id: string;
  title: string;
  metric_type: string;
  target_value?: number;
  current_value?: number;
  confidence?: number;
  status: string;
};

export type OKRObjective = {
  id: string;
  cycle_id: string;
  team_id?: string;
  owner_user_id?: string;
  title: string;
  description?: string;
  status: string;
  tags?: string[];
};

export type OKRLink = {
  id: string;
  objective_id?: string;
  key_result_id?: string;
  entity_type: string;
  entity_id: string;
  relation: string;
};

export type OKRCycle = {
  id: string;
  name: string;
  type: string;
  team_id?: string;
  start_date?: string;
  end_date?: string;
  status: string;
};

export type OKRCoverage = {
  entity_type: string;
  coverage: number;
  linked_key_results: number;
  total_key_results: number;
  deviation: number;
};

export type OKRMetrics = {
  completion: number;
  deviation: number;
  unlinked_key_results: OKRKeyResult[];
  coverage: OKRCoverage[];
};

export type OKRFilter = {
  cycle_id?: string;
  team_id?: string;
  owner_user_id?: string;
  status?: string;
};

export async function listCycles() {
  const { data } = await api.get("/okrs/cycles");
  return data.data as OKRCycle[];
}

export async function createCycle(payload: {
  name: string;
  type?: string;
  status?: string;
  start_date?: string;
  end_date?: string;
  team_id?: string;
}) {
  const { data } = await api.post("/okrs/cycles", payload);
  return data.data as OKRCycle;
}

export async function updateCycle(id: string, payload: Partial<OKRCycle>) {
  const { data } = await api.put(`/okrs/cycles/${id}`, payload);
  return data.data as OKRCycle;
}

export async function listObjectives(filter: OKRFilter) {
  const { data } = await api.get("/okrs/objectives", { params: filter });
  return data.data as { objectives: OKRObjective[]; key_results: OKRKeyResult[] };
}

export async function createObjective(payload: {
  cycle_id: string;
  title: string;
  description?: string;
  status?: string;
  team_id?: string;
  owner_user_id?: string;
  tags?: string[];
}) {
  const { data } = await api.post("/okrs/objectives", payload);
  return data.data as OKRObjective;
}

export async function updateKRProgress(id: string, current: number, confidence: number) {
  return api.post(`/okrs/key-results/${id}/progress`, { current, confidence });
}

export async function createKeyResult(payload: {
  objective_id: string;
  title: string;
  metric_type?: string;
  target_value?: number;
  unit?: string;
  current_value?: number;
  confidence?: number;
  status?: string;
}) {
  const { data } = await api.post("/okrs/key-results", payload);
  return data.data as OKRKeyResult;
}

export async function addLink(link: Omit<OKRLink, "id">) {
  return api.post("/okrs/links", link);
}

export async function removeLink(id: string) {
  return api.delete(`/okrs/links/${id}`);
}

export async function fetchMetrics(filter: OKRFilter) {
  const { data } = await api.get("/okrs/metrics", { params: filter });
  return data.data as OKRMetrics;
}

File: internal/models/okr.go
package models

import (
	"time"

	"github.com/google/uuid"
	"gorm.io/datatypes"
)

type OKRCycle struct {
	ID        uuid.UUID  `gorm:"type:uuid;primaryKey;default:gen_random_uuid()"`
	OrgID     uuid.UUID  `gorm:"type:uuid;index"`
	TeamID    *uuid.UUID `gorm:"type:uuid;index"`
	Type      string
	Name      string
	StartDate time.Time `gorm:"type:date;column:start_date"`
	EndDate   time.Time `gorm:"type:date;column:end_date"`
	Status    string
	CreatedBy uuid.UUID `gorm:"type:uuid;column:created_by"`
	Timestamps
	SoftDelete
}

func (OKRCycle) TableName() string { return "okr_cycles" }

type Objective struct {
	ID                uuid.UUID  `gorm:"type:uuid;primaryKey;default:gen_random_uuid()"`
	OrgID             uuid.UUID  `gorm:"type:uuid;index"`
	CycleID           uuid.UUID  `gorm:"type:uuid;index;column:cycle_id"`
	TeamID            *uuid.UUID `gorm:"type:uuid;index"`
	OwnerUserID       uuid.UUID  `gorm:"type:uuid;index;column:owner_user_id"`
	ParentObjectiveID *uuid.UUID `gorm:"type:uuid;index;column:parent_objective_id"`
	Title             string
	Description       *string
	Weight            float64
	Status            string
	Tags              datatypes.JSON `gorm:"type:jsonb"`
	SchemaVersion     int            `gorm:"column:schema_version"`
	Payload           datatypes.JSON `gorm:"type:jsonb"`
	Timestamps
	SoftDelete
}

func (Objective) TableName() string { return "objectives" }

type KeyResult struct {
	ID           uuid.UUID `gorm:"type:uuid;primaryKey;default:gen_random_uuid()"`
	OrgID        uuid.UUID `gorm:"type:uuid;index"`
	ObjectiveID  uuid.UUID `gorm:"type:uuid;index;column:objective_id"`
	Title        string
	MetricType   string `gorm:"column:metric_type"`
	TargetValue  *float64 `gorm:"column:target_value"`
	CurrentValue *float64 `gorm:"column:current_value"`
	Unit         *string
	Weight       float64
	Confidence   int
	Status       string
	SchemaVersion int `gorm:"column:schema_version"`
	MetricPayload datatypes.JSON `gorm:"type:jsonb;column:metric_payload"`
	Timestamps
	SoftDelete
}

func (KeyResult) TableName() string { return "key_results" }
